### ISSUE TITLE
Duplicate parent_left / parent_right values

### ISSUE THREAD

**Issue by mcb30** [odoo/odoo#14493]
Impacted versions:

All

Steps to reproduce:

Create a new top-level package in each of two concurrent transactions, then commit both transactions.

Current behavior:

We are finding that Odoo is occasionally creating duplicate parent_left/parent_right values in stock.quant.package, which breaks the nested set model.

We suspect that this problem is arising from a race condition in which two concurrent transactions both create a new package, and both end up with the same values for parent_left/parent_right.  There seems to be no code in odoo/models.py's _create() that would prevent this from happening.

Expected behavior:

parent_left and parent_right values should always be unique within a table.

Video/Screenshot link (optional):

Here is some example data from our production system illustrating the problem:

`SELECT parent_left FROM stock_quant_package GROUP BY parent_left HAVING COUNT(1) > 1; SELECT parent_right FROM stock_quant_package GROUP BY parent_right HAVING COUNT(1) > 1;`

 parent_left
-------------
       17062
       17063
       17204
       17379
(4 rows)

 parent_right
--------------
        17064
        17091
        17380
(3 rows)

`SELECT id, parent_id, name, parent_left, parent_right, create_date FROM stock_quant_package WHERE parent_left >= 17062 AND parent_right <= 17091 ORDER BY parent_id NULLS FIRST, id;`

  id  | parent_id |     name     | parent_left | parent_right | create_date
------|-----------|--------------|-------------|--------------|---------------------------- 
 8777 |           | PAL000007017 |       17062 |        17091 | 2016-11-28 09:05:28.806007
 8779 |           | PAL000005672 |       17062 |        17091 | 2016-11-28 09:05:29.643267
 8778 |      8777 | PACK0007857  |       17063 |        17064 | 2016-11-28 09:05:28.806007
 8780 |      8779 | PACK0007859  |       17063 |        17064 | 2016-11-28 09:05:29.643267
 8802 |      8779 | PACK0007881  |       17065 |        17066 | 2016-11-28 09:23:40.229422
 8803 |      8779 | PACK0007882  |       17067 |        17068 | 2016-11-28 09:24:30.892247
 8804 |      8779 | PACK0007883  |       17069 |        17070 | 2016-11-28 09:25:45.361386
 8805 |      8779 | PACK0007884  |       17071 |        17072 | 2016-11-28 09:33:54.093324
 8806 |      8779 | PACK0007885  |       17073 |        17074 | 2016-11-28 09:34:29.544915
 8807 |      8779 | PACK0007886  |       17075 |        17076 | 2016-11-28 09:35:13.6721
 8808 |      8779 | PACK0007887  |       17077 |        17078 | 2016-11-28 09:36:30.484778
 8809 |      8779 | PACK0007888  |       17079 |        17080 | 2016-11-28 09:37:03.190428
 8810 |      8779 | PACK0007889  |       17081 |        17082 | 2016-11-28 09:37:38.886488
 8811 |      8779 | PACK0007890  |       17083 |        17084 | 2016-11-28 09:39:01.968425
 8814 |      8779 | PACK0007893  |       17085 |        17086 | 2016-11-28 09:39:35.076745
 8818 |      8779 | PACK0007897  |       17087 |        17088 | 2016-11-28 09:40:35.086901
 8857 |      8779 | PACK0007927  |       17089 |        17090 | 2016-11-28 10:07:15.216961
(17 rows)

Additional questions:

Should there be a UNIQUE constraint on each of the parent_left and parent_right columns?  This would at least prevent data corruption.

How can the code in models.py be fixed to prevent this problem?  I would expect that the logic in _create() and _write() should be using FOR UPDATE to lock the appropriate rows, but I have not yet examined the transactions in detail. 

Thanks,

Michael

**Comment by mcb30** [comment]
@rco-odoo: I suspect this issue may require your attention.

**Comment by rco-odoo** [comment]
@odony can you have a look at this?

@mcb30 from what I understand, both transactions insert a new row in the table and then update the columns `parent_left` and `parent_right` in their respective row. As the new records have no parent, we should fall in this case https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L3926.

The last `UPDATE` query from the first transaction modifies `parent_right`, which has an impact on the `SELECT MAX(parent_right)` from the second transaction. Therefore Postgres should not allow the second transaction to commit, right? Otherwise, we violate the principle of "repeatable read" in transactions. If the transactions use the isolation level "serializable" (which is the default for Odoo), this should not happen.

**Comment by mcb30** [comment]
On 28/11/16 15:28, Raphael Collet wrote:
> @mcb30 <https://github.com/mcb30> from what I understand, both
> transactions insert a new row in the table and then update the columns
> |parent_left| and |parent_right| in their respective row. As the new
> records have no parent, we should fall in this case
> https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L3926.
>
> The last |UPDATE| query from the first transaction modifies
> |parent_right|, which has an impact on the |SELECT MAX(parent_right)|
> from the second transaction. Therefore Postgres should not allow the
> second transaction to commit, right? Otherwise, we violate the principle
> of "repeatable read" in transactions. If the transactions use the
> isolation level "serializable" (which is the default for Odoo), this
> should not happen.

It looks as though Odoo never uses "serializable":

   https://github.com/odoo/odoo/blob/10.0/odoo/sql_db.py#L317

In the "repeatable read" isolation level, which it seems that Odoo uses, 
there will be no predicate lock as a result of the MAX(parent_right) and 
so nothing to prevent the second transaction from committing.

Michael

**Comment by rco-odoo** [comment]
@mcb30 you are right, we use "repeatable read" instead of "serializable" for performance reasons.

Together with @odony we have found a possible way to prevent concurrent inserts inside the table, which is the origin of the bug. Before the three UPDATE queries at https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L3931, add
```python
   # lock the table to prevent other rows to use the same parent_left or parent_right
   cr.execute("LOCK TABLE %s IN SHARE MODE" % self._table)
```

We have tested it by hand on a database, and it seems to work as expected. Once a row is inserted or modified, then either the reader transactions or the writer transaction of the table are aborted.

Can you give it a try and confirm the problem goes away?
I will prepare a fix for versions 8.0 and later.

Thank you very much!
Raphael

**Comment by odony** [comment]
@mcb30 In addition, we agree that a UNIQUE constraint/index on `parent_left` and `parent_right` is necessary, but we need to guard it with the explicit locking as shown by @rco-odoo. Otherwise when the race condition occurs one of the transactions will crash instead of being automatically retried as it should.

**Comment by mcb30** [comment]
On 28/11/16 16:25, Raphael Collet wrote:
> @mcb30 <https://github.com/mcb30> you are right, we use "repeatable
> read" instead of "serializable" for performance reasons.
>
> Together with @odony <https://github.com/odony> we have found a possible
> way to prevent concurrent inserts inside the table, which is the origin
> of the bug. Before the three UPDATE queries at
> https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L3931, add
>
>     # lock the table to prevent other rows to use the same parent_left or parent_right
>     cr.execute("LOCK TABLE%s  IN SHARE MODE"  %  self._table)

That should work (assuming that the code goes in both _create() and 
_write()), but will block any other updates to the whole table until the 
transaction completes, including updates that do not modify 
parent_left/parent_right.  Is there a viable solution that is less 
disruptive to other transactions?

Thanks,

Michael

**Comment by rco-odoo** [comment]
@mcb30 indeed, the statement `LOCK TABLE ... IN SHARE MODE` prevents any concurrent update in the given table. Ideally you would like to lock only the columns `parent_left` and `parent_right`, but in PostgreSQL you can only lock rows or the whole table. So this will indeed restrict concurrent creations on models with the `parent_left`/`parent_right` trick.

Note that it is actually enought to put the `LOCK` statement in method `_create` only. Concurrent calls to `_create` or `_write` will protect against each other in all cases, because `INSERT` and `UPDATE` statements will lock individual rows with a lock that conflicts with the `SHARE` lock on the table:

Transactions | _create | _write
---|---|---
_create | LOCK (_create) >< INSERT (_create) | LOCK (_create) >< UPDATE (_write)
_write  | UPDATE (_write) >< LOCK (_create) | UPDATE (_write) >< UPDATE (_write)

Note that two concurrent calls to `_write` will only conflict if the updated rows overlap. This allows concurrent updates on disjoint sets of rows.

Cheers,
Raphael

**Comment by odony** [comment]
@mcb30 The patch will indeed block other updates on the table, but in real life this should rarely happen. We typically only use the "parent storing" option on heavily-nested hierarchical master data, such as stock locations. Such master data is rarely updated/inserted, ands it's supposed to be done only by the person in charge, so it should not be done concurrently (even when imported in batch).
This is probably the reason while the problem was never reported before.

There is usually no point storing the parent structure for operational data where only a single subtree needs to be accessed at a time, therefore without much nesting. It may have been a mistake to activate this option on `stock_quant_package`, and a simple fix might be to drop the `_parent_store` option entirely and switch back to explicit tree traversal. Such a fix would be rather invasive for a stable branch, but might still be considered as an extra performance patch if the explicit LOCKING solution causes performance issues in the case of `stock_quant_package`.

**Comment by rco-odoo** [comment]
@mcb30 @odony after deeper testing and analysis, it turns out that the locking solution does **not** fix the issue: I was able to run two transactions that create records with the same (parent_left, parent_right). The lock only reduces the probability of the issue to occur, but it does not prevent it.

I am now considering a unicity constraint on the columns `parent_left`, `parent_right`. It will for sure prevent the inconsistencies you have observed. The tricky part is to turn the constraint violation into a concurrent access error, so that the request is retried instead of failing for an obscure reason.

**Comment by rco-odoo** [comment]
@mcb30 I finally found an effective way to constrain the values of `parent_left`, `parent_right` to avoid the situation you experienced. The pull request #14562 proposes a fix to the race condition. It adds a constraint that, when violated, is considered as a concurrent update error rather than a traditional constraint violation.

Could you give a try to #14562 and confirm whether it fixes your case? If it does, we will merge this one to version 10.0.

Concurrently to this, we will remove the `parent_left`, `parent_right` stuff from the model `stock.quant.package` for the next version.

**Comment by florian-dacosta** [comment]
I face the same problem with version 9.
I'd like to know if you plan to fix this issue for previous version ? (8, 9, ...)
If you have a PR for version 9, I can reproduce the bug easily and test it.

**Comment by mcb30** [comment]
On 16/02/17 14:42, Florian wrote:
> I face the same problem with version 9.
> I'd like to know if you plan to fix this issue for previous version ?
> (8, 9, ...)
> If you have a PR for version 9, I can reproduce the bug easily and test it.

If it helps: we worked around it by disabling the nested set model for 
stock.quant.package via an addon:

   from odoo import api, fields, models
   class QuantPackage(models.Model):
       _inherit = 'stock.quant.package'
       _parent_store = False

Michael

**Comment by florian-dacosta** [comment]
@mcb30 Thank you for the advice, I already did it for my customer.
But I think it should really be fixed in core!

**Comment by sylvain-garancher** [comment]
We also have this issue.

I just tried it on a test database to set `_parent_store = False`, and it seems to work... But this doesn't.
Doing this will not the columns from the database, and the `child_of` operator still uses these columns, which will never be updated again.

After I really remove the `parent_left` and `parent_right` columns from the database, using the `child_of` operator in a search results in an SQL error `ProgrammingError: column stock_quant_package.parent_left does not exist`.

So we also need to remove all occurences of the `child_of` operator in searches for `stock.quand.package`.
Luckily, there are only three of them in the standard code:
- https://github.com/odoo/odoo/blob/2cd63aa2dc2ab0706922a660242b2132694e41a8/addons/stock/models/stock_quant.py#L464
- https://github.com/odoo/odoo/blob/2cd63aa2dc2ab0706922a660242b2132694e41a8/addons/stock/models/stock_quant.py#L623
- https://github.com/odoo/odoo/blob/2cd63aa2dc2ab0706922a660242b2132694e41a8/addons/stock/models/stock_quant.py#L713

**Comment by nim-odoo** [comment]
I guess the fields have been removed from the database, but they are sill defined at the model level, right? If you are about to remove the fields anyway, you might simply apply the complete patch: https://github.com/odoo/odoo/pull/14694

**Comment by odony** [comment]
@sylvain-garancher setting `_parent_store` to False should be sufficient as a workaround for an existing database. The framework [can handle](https://github.com/odoo/odoo/blob/8749410b1033ddec1207ce1db42d1889a0d2ea33/openerp/osv/expression.py#L718) `child_of` operator on models where `_parent_store` is `False`, it will simply do multiple queries for traversing the hierarchy. The `parent_left` and `parent_right` fields will also be ignored if they are present, so you do not need to change them.

What kind of abnormal behavior did you see when you simply used the workaround proposed in https://github.com/odoo/odoo/issues/14493#issuecomment-280715903 ?

**Comment by sylvain-garancher** [comment]
I wanted to check that the fields were not used anymore when calling search with the child_of operator, but I forgot to remove them from the python, as @nim-odoo said :)

I just retrid with the full patch, and I confirm that it works as expected.
Sorry for the noise.

**Comment by alnslang** [comment]
@nim-odoo and @odony Is-it possible to merge this patch in v10 ?

**Comment by odony** [comment]
@alnslang @nim-odoo yes, we could consider turning off the _parent_store feature on `stock.quant.package` in 10.0, as the system should gracefully fallback to recursing through the hierarchy.
It's not very pretty on the "stable" scale, but I don't think it should cause any issue, and should not affect performance negatively, given the depth of typical "package" hierarchies. Good testing of the seamless "update" (without using `-u stock`) on existing databases is recommended, though.

**Comment by nim-odoo** [comment]
@odony 
Well, what surprises me is that removing the field from the DB caused an error:
> After I really remove the parent_left and parent_right columns from the database, using the child_of operator in a search results in an SQL error ProgrammingError: column stock_quant_package.parent_left does not exist.

I would have thought that if `_parent_store = False`, this shouldn't happen. I would have expected that in this case, we never explicitly require this field to exist in the DB. I will dig into it.

**Comment by odony** [comment]
@nim-odoo `parent_{left,right}` are `integer` fields so they are prefetched

**Comment by pedrobaeza** [comment]
@odony can this be closed?

**Comment by odony** [comment]
This specific problem is gone as of 11.0 thanks to #14694, and the root cause is fixed in master thanks to #22558, so it can be closed indeed.

**Comment by vumgg** [comment]
✌️
